modules/create_timepoint/jsx/create_timepoint.js

```diff
@@ -0,0 +1,352 @@
+        },
+        body: send,
+      }
+    ).then((response) => response.json())
+      .then(
+        (data) => {
+          // Populate the form errors.
+          if (data.errors && data.errors.length > 0) {
+            this.setState({errors: data.errors});
+          }
+          // Populate the select options for subproject.
+          if (data.subproject) {
+            const state = Object.assign({}, this.state);
+            state.form.options.subproject = data.subproject;
+            state.form.value.subproject = Object.keys(data.subproject)[0];
+            state.form.display.subproject = true;
+            this.setState(state);
+          }
+          // Populate the select options for psc.
+          if (data.psc) {
+            const state = Object.assign({}, this.state);
+            state.form.options.psc = data.psc;
+            state.form.value.psc = Object.keys(data.psc)[0];
+            state.form.display.psc = true;
+            this.setState(state);
+          }
+          // Populate the select options for visit.
+          if (data.visit) {
+            // Store the (complete) visit selection information.
+            const state = Object.assign({}, this.state);
+            state.storage.visit = data.visit;
+            this.setState(state);
+            // Handle visit selection.
+            this.handleVisitLabel();
+          }
+          // Display form to user.
+          this.setState({isLoaded: true});
+        }).catch((error) => {
+          this.populateErrors({message: 'Server error.'});
+          this.setState({isLoaded: true});
+        });
+  }
+  /**
+   * Visit Labels refreshes when Subproject changes.
+   */
+  handleVisitLabel() {
+    const state = Object.assign({}, this.state);
+    state.form.options.visit = state.storage.visit[
+      state.form.value.subproject
+    ];
+    state.form.value.visit = Object.keys(state.storage.visit[
+      state.form.value.subproject
+    ])[0];
+    state.form.display.visit = true;
+    this.setState(state);
+  }
+  /**
+   * Set the form data based on state values of child elements/components
+   *
+   * @param {string} formElement - name of the selected element
+   * @param {string} value - selected value for corresponding form element
+   */
+  setForm(formElement, value) {
+    const state = Object.assign({}, this.state);
+    state.form.value[formElement] = value;
+    this.setState(state);
+    if (formElement === 'subproject') {
+      this.handleVisitLabel();
+    }
+  }
+
+  /**
+   * Populate the elements of errors to display.
+   *
+   * @param {object} values - for individual form element.
+   */
+  populateErrors(values) {
+    let errors = [];
+    Object.keys(values).forEach(function(key) {
+      errors.push(
+        <div className='col-xs-12 col-sm-12 col-md-12'>
+          <label className='error form-group'>
+            {values[key]}
+          </label>
+        </div>
+      );
+    });
+    this.setState({errors: errors});
```

Comment:
what data type is this.state.errors? it seems weird to me that it's being set here as an array of jsx. they seem to be strings elsewhere in loris

Reply:
The errors should be just a string from the server.

Reply:
isn't it an array of jsx here however?

Reply:
@zaliqarosli Oh yes you're correct. Should I rename errors to something else. I guess multiple errors could be returned and were stacked on top of each other. Should I name errors to something else?

Reply:
the issue here is that this.state.errors is defined as a boolean (false), and then set as a string from 'data.errors' (line 123), and then set as an array of jsx in populateErrors() (line 202). i think this.state.errors should either be a boolean (true or false that there is an error) or a string (the error message itself), and any rendering of the error is instead done outside of the state. does this make sense?

Reply:
Hey @zaliqarosli is it alright if I make the state.errors be null and it becomes the divs with error messages to be shown when errors are returned from the server. Please see new code changes. I think it's simpler this way and I don't know if rendering of the error would be better if done outside of the state.

Reply:
i think its still best practice to put in state the most minimal amount of data needed to represent your UI's state, leaving all render props to be in the render function. this way, the rendering of the error is only defined once, and only the error data gets re-rendered on change. because of this, i would argue that rendering of the error is definitely better done inside the render function (it is what it's made for), and not as data passed from the state object

Reply:
@zaliqarosli I'm a little confused. The error(s) from the server are stored in the state and so it will re-render the change of the errors to display in the render function. The error(s) can change depending on the user submitting the form submission and the depending on the error(s)-case in question. I'm using the state to trigger when things are to be displayed as well. I guess since I'm storing possibly multiple divs (depending on single error or multiple) is what's upsetting. That's sort of the reason why it ended up the way it did. Are you suggesting I place the function in render and have it be triggered by an array that's stored in state?

Reply:
im not sure i fully understand but what I am trying to say is that (referring to lines 196-204) we shouldn't be storing jsx elements in state. if you are trying to display multiple errors, then the errors prop should be an array of string errors. how this array of errors is rendered should be decided inside a render function. so perhaps the div of errors should be inserted on line 305. 

in other words, you are storing divs of error labels in state, and then rendering that state data. what you should be doing is storing an array of string errors in state, and then rendering divs of labels of that state data.

Reply:
@zaliqarosli I added the change to remove jsx from state.